From a17f78ef069c1d2dede79a1068823b97bca4bcb7 Mon Sep 17 00:00:00 2001
From: Robert Marko <robimarko@gmail.com>
Date: Thu, 15 Dec 2022 22:33:27 +0100
Subject: [PATCH] ath11k: support smallmem variant

ath11k is really memory intensive for devices with less that 1GB of RAM,
so lets port provide a smallmem variant for devices with 512MB of RAM.

It saves a significant amount of memory by setting the FW to Mode-1
however the drawback is reduced number of VDEV-s and peers.

Currently, only AHB devices (IPQ8074 and IPQ6018) have a smallmem
profile but it can be extended for PCI devices as well.

Signed-off-by: Robert Marko <robimarko@gmail.com>
---
 drivers/net/wireless/ath/ath11k/core.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -83,9 +83,15 @@ static const struct ath11k_hw_params ath
 		.supports_sta_ps = false,
 		.cold_boot_calib = false,
 		.cbcal_restart_fw = true,
+#ifndef CONFIG_ATH11K_SMALLMEM
 		.fw_mem_mode = 0,
 		.num_vdevs = 16 + 1,
 		.num_peers = 512,
+#else
+		.fw_mem_mode = 1,
+		.num_vdevs = 8,
+		.num_peers = 128,
+#endif
 		.supports_suspend = false,
 		.hal_desc_sz = sizeof(struct hal_rx_desc_ipq8074),
 		.supports_regdb = false,
@@ -163,9 +169,15 @@ static const struct ath11k_hw_params ath
 		.supports_sta_ps = false,
 		.cold_boot_calib = true,
 		.cbcal_restart_fw = true,
+#ifndef CONFIG_ATH11K_SMALLMEM
 		.fw_mem_mode = 0,
 		.num_vdevs = 16 + 1,
 		.num_peers = 512,
+#else
+		.fw_mem_mode = 1,
+		.num_vdevs = 8,
+		.num_peers = 128,
+#endif
 		.supports_suspend = false,
 		.hal_desc_sz = sizeof(struct hal_rx_desc_ipq8074),
 		.supports_regdb = false,
